controllers/secretdefinition_controller.go

```diff
@@ -0,0 +1,242 @@
+	secret := &corev1.Secret{
+		Type: corev1.SecretType(secretDef.Spec.Type),
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: secretDef.Namespace,
+			Labels: map[string]string{
+				"managedBy":     "secrets-manager",
+				"lastUpdatedAt": time.Now().Format(timestampFormat),
+			},
+			Name: secretDef.Spec.Name,
+		},
+		Data: data,
+	}
+	err := r.Create(r.Ctx, secret)
+	if errors.IsAlreadyExists(err) {
+		err = r.Update(r.Ctx, secret)
+	}
+	return err
+}
+
+// deleteSecret will delete a secret given its namespace and name
+func (r *SecretDefinitionReconciler) deleteSecret(namespace string, name string) error {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+	}
+	return r.Delete(r.Ctx, secret)
+}
+
+// +kubebuilder:rbac:groups=secrets-manager.tuenti.io,resources=secretdefinitions,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=secrets-manager.tuenti.io,resources=secretdefinitions/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;patch;delete
+func (r *SecretDefinitionReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("secretdefinition", req.NamespacedName)
+
+	secretDef := &secretsmanagerv1alpha1.SecretDefinition{}
+
+	err := r.Get(r.Ctx, req.NamespacedName, secretDef)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("could not get SecretDefinition '%s'", req.NamespacedName))
+		return ctrl.Result{}, ignoreNotFoundError(err)
+	}
+
+	secretName := secretDef.Spec.Name
+	secretNamespace := secretDef.Namespace
+
+	log = log.WithValues("secret", fmt.Sprintf("%s/%s", secretNamespace, secretName))
+
+	if secretDef.ObjectMeta.DeletionTimestamp.IsZero() {
+
+		// Let's add the finalizer if it's not present
+		if !containsString(secretDef.ObjectMeta.Finalizers, finalizerName) {
+			secretDef.ObjectMeta.Finalizers = append(secretDef.ObjectMeta.Finalizers, finalizerName)
+			if err = r.Update(r.Ctx, secretDef); err != nil {
+				log.Error(err, "unable to update SecretDefinition finalizers", "finalizer", finalizerName)
+				return ctrl.Result{}, err
+			}
+		}
+
+		// Get data from the secret source of truth
+		desiredState, err := r.getDesiredState(secretDef.Spec.KeysMap)
+
+		if err != nil {
+			log.Error(err, "unable to get desired state for secret")
+			secretSyncErrorsTotal.WithLabelValues(secretName, secretNamespace).Inc()
+			secretLastSyncStatus.WithLabelValues(secretName, secretNamespace).Set(0.0)
+			return ctrl.Result{}, err
+		}
+
+		// Get the actual secret from Kubernetes
+		currentState, err := r.getCurrentState(secretNamespace, secretName)
+
+		if err != nil && !errors.IsNotFound(err) {
+			log.Error(err, "unable to get current state of secret")
+			secretSyncErrorsTotal.WithLabelValues(secretName, secretNamespace).Inc()
```

Comment:
(👿  is in the details)  across the code, the pairing for locate a secret is a "tuple" `namespace, secretName`, but in this point this is fliped: (`secretName, namespace`) 

ultra minor thing I know... but since this codebase is really really "mimado" I'd use always the same order for locating a secret :)

Reply:
good catch! makes sense!!